text/1004-transactions-api.rst

```diff
@@ -651,16 +945,129 @@ There are few problems of this approach:
+and turn the whole block of code into a transaction rather than just a
+mutation. I.e. retrying a single mutable request on a "concurrent
+update" error must be a deliberate decision.
+
+
+Is ``with_modifications`` Needed?
+---------------------------------
+
+It's intuitive that databases are mutable by default. But there is a
+large class of applications that are mostly read-only and must have
+limited and easy to find places having mutations. For those apps it's
+better to have read-only connection by default and use a pattern like
+this for writes:
+
+.. code-block:: python
+
+    def save(conn):
+        async for tx in conn.with_mutations().retry():
+          async with tx:
+            ...
+
+While retry is also easy to find, nothing stops user from writing:
+
+.. code-block:: python
+
+    conn.fetch("INSERT User { .. }")
+
+Except the read_only configuration.
+
+The metadata is will probably be implemented later and works the same:
+
+.. code-block:: python
+
+    def http_handler(req, conn):
+        req['conn'] = conn.with_metadata(url=req.url)
+        process_request(req)
+
+Use Context Manager
+-------------------
+
+While it's tempting to use context manager for configuring connection,
+in particular for the session config:
+
+.. code-block:: python
+
+    def handler(conn):
+        with conn.session_config(name="value"):
+            conn.query("...")
+
+But behavior implies that session config is reset on each
+``with/__exit__``, which has consequences:
+
+* When ``handler()`` is interrupted, e.g. a timeout occurred, connection
+  could be left in the inconsistent state (i.e. should be dropped)
+* It prevents caching the session config, e.g. in the case above:
+  ``handler(c); handler(c)`` will not set session config second
+  time, but use the knowledge that these settings are already active.
+
+Also it improves clarity and composition. E.g. (here we demonstrate
+``with_metadata`` method with is a part of `Future Work`_:
+
+.. code-block:: python
+
+    def money_transfer(tx):
+        user1 = tx.with_metadata(user='user1')
+        user2 = tx.with_metadata(user='user2')
+        withdraw(user1)
+        deposit(user2)
+        log_withdraw(user1)
+        log_deposit(user2)
+
+(This implies ``Transaction`` also contains ``with_metadata`` method)
+
+Additional thing considered is that with context manager, if connection
+is failed in the middle of the block:
+
+.. code-block:: python
+
+    with conn.session_config(name="value"):
+        conn.query("...")
+        # << -- here
+        conn.query("...")
+
+The implementation should reconnect and reapply the session
+configuration anyway. So it deviates from the usual behavior of context
+manager anyway.
+
+
+Disabling Features
+------------------
+
+We may introduce a pool and/or connection configuration to disable
+``ddl`` and ``config`` features on the requests. And/or disable
+``session`` and ``transaction`` features on connection. This should be
```

Comment:
I like the disabling idea. I'm not sure if it should be the default configuration though, we can just teach that it's a good pattern to narrow down the connection functionality to what your app needs.

Reply:
Agree.  I think we should optimize for simplicity at a glance, in which case lengthy connection "setups" in simple tutorials would scare users away.

Reply:
Looks like you agree with what is written in the RFC. Resolving the conversation.